arch/risc-v/src/bl602/bl602_flash.c

```diff
@@ -0,0 +1,103 @@
+/****************************************************************************
+ * arch/risc-v/src/bl602/bl602_flash.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdint.h>
+#include <syslog.h>
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct bl602_romflash_cfg_desc
+{
+  uint32_t magic;
+  uint8_t  cfg[84];
+};
+
+#define ROMAPI_BASE                             (0x21010800)
```

Comment:
@liangzhanggb does ROM function handle the sync(disable/enable interrupt) in ROM function? If not, either method is fine for me.

Reply:
arch/risc-v/src/bl602/bl602_flash.c
`
ROMAPI_SFLASH_EREASE_NOLOCK
ROMAPI_SFLASH_WRITE_NOLOCK
ROMAPI_SFLASH_READ_NOLOCK
ROMAPI_SFLASH_EREASE_WITHLOCK
ROMAPI_SFLASH_WRITE_WITHLOCK
ROMAPI_SFLASH_READ_WITHLOCK
...
`

We have two kinds of romapi. The romapi ending with "_NOLOCK" has "dis/enable interrupt", and the romapi ending with "_WITHLOCK" has no "dis/enable interrupt".

Reply:
Does it actually do that with interrupts?  From my ROM dump it looked to me like those functions were just providing ownership of the SFlash where the difference is the with lock functions would call `XIP_SFlash_State_Save` and `XIP_SFlash_State_Restore`.  I did not see any logic that would touch the `CLIC`.  This is probably out of the scope of this PR, but something we should be aware of.

Reply:
arch/risc-v/src/bl602/bl602_flash.c

1. 
ROMAPI_SFLASH_EREASE_NEEDLOCK
ROMAPI_SFLASH_WRITE_NEEDLOCK
ROMAPI_SFLASH_READ_NEEDLOCK

These functions only provide ownership of SFlash. 

I have corrected it with following way
{
  flags = up_irq_save();
  *_NEEDLOCK(...)
  up_irq_restore(flags);
}


2. 
ROMAPI_SFLASH_EREASE_WITHLOCK
ROMAPI_SFLASH_WRITE_WITHLOCK
ROMAPI_SFLASH_READ_WITHLOCK
ROMAPI_SFLASH_EREASE_WITHLOACK =
{
  disable_irq(...)
  *_NEEDLOCK(...)
  enable_irq(...)
}

Reply:
@xiaoxiang781216 @liangzhanggb  I am thinking we should just not bother with disabling the IRQs here for now and later address this.  Doing a long flash write without interrupts enabled will cause everything to lock which would be really bad for performance even though we are only trying to protect register writes.

Reply:
@btashton Our code run through the sflash controller. when operating the sflash controller, it is not allowed to run the code (from flash). Unless the code is copied to ram, But "IoT chip"s rarely do this because there is very little memory. So we disable interrupts here. Thank you very much.

Reply:
Ah ok. This makes more sense. Let's go with what you have.